src/main/scala/components/csr.scala

```diff
@@ -0,0 +1,533 @@
+    MCSRs.minstret -> reg_instret,
+    MCSRs.mimpid -> 0.U,
+    MCSRs.marchid -> 0.U,
+    MCSRs.mvendorid -> 0.U,
+    MCSRs.misa -> reg_misa.asUInt,
+    MCSRs.mstatus -> read_mstatus,
+    MCSRs.mtvec -> MCSRCmd.TRAPADDR,
+    MCSRs.mip -> reg_mip.asUInt(),
+    MCSRs.mie -> reg_mie.asUInt(),
+    MCSRs.mscratch -> reg_mscratch,
+    MCSRs.mepc -> reg_mepc,
+    MCSRs.mtval -> reg_mtval,
+    MCSRs.mcause -> reg_mcause.asUInt(),
+    MCSRs.mhartid -> 0.U,
+    MCSRs.medeleg -> reg_medeleg)
+    
+  read_mapping += MCSRs.mcycleh -> 0.U 
+  read_mapping += MCSRs.minstreth -> 0.U
+  
+  //CSR DECODE
+  val cmd = WireInit(3.U(3.W))
+  
+  when( io.decode.inst(6, 0) === ("b1110011".U)){
+    switch(io.decode.inst(14, 12)){
+      is("b011".U){
+        cmd := MCSRCmd.clear
+        io.regwrite := true.B
+      }
+      is("b111".U){
+        cmd := MCSRCmd.clear
+        io.regwrite := true.B
+      } 
+      is("b010".U){
+        cmd := MCSRCmd.set
+        io.regwrite := true.B
+      }
+      is("b110".U){
+        cmd := MCSRCmd.set
+        io.regwrite := true.B
+      }
+      is("b001".U){
+        cmd := MCSRCmd.write
+        io.regwrite := true.B
+      }
+      is("b101".U){
+        cmd := MCSRCmd.write
+        io.regwrite := true.B
+      }
+      is("b000".U){
+        cmd := MCSRCmd.interrupt
+        io.regwrite := false.B
+      }
+    }
+  }.otherwise{
+    cmd := MCSRCmd.nop
+    io.regwrite := false.B
+  }
+  
+  val csr = io.decode.inst(MCSRCmd.MSB, MCSRCmd.LSB)
+  val system_insn = cmd === MCSRCmd.interrupt
+  val cpu_ren = cmd =/= MCSRCmd.nop && !system_insn
+
+
+  val decoded_addr = read_mapping map { case (k, v) => k -> (csr === k) }
+  val priv_sufficient = MCSRCmd.MPRV >= csr(9,8)
+  val read_only = csr(11,10).andR
+  val cpu_wen = cpu_ren && cmd =/= MCSRCmd.read && priv_sufficient
+  val wen = cpu_wen && !read_only
+  val wdata = readModifyWriteCSR(cmd.asInstanceOf[UInt], io.rw.rdata, io.rw.wdata)
+
+  val opcode = 1.U << csr(2,0)
+  val insn_call = system_insn && opcode(0)
+  val insn_break = system_insn && opcode(1)
+  val insn_ret = system_insn && opcode(2) && priv_sufficient
+  val insn_wfi = system_insn && opcode(5) && priv_sufficient
+
+  private def decodeAny(m: LinkedHashMap[Int,Bits]): Bool = m.map( { case(k: Int, _: Bits) => csr === k }).reduce(_ || _)
+  io.decode.read_illegal := 3 < csr(9,8) || !decodeAny(read_mapping) 
+  io.decode.write_illegal := csr(11,10).andR
+  io.decode.system_illegal := 3 < csr(9,8)
+
+  io.status := reg_mstatus
+
+  io.eret := insn_call || insn_break || insn_ret
+
+  // ILLEGAL INSTR
+  when (io.exception) {
+    reg_mcause.interrupt := MCauses.illegal_instruction & "h80000000".U
+    reg_mcause.exceptioncode := MCauses.illegal_instruction & "h7fffffff".U
+    io.evec := "h80000004".U
+    reg_mepc := io.pc // misaligned memory exceptions not supported...
+  }
+
+  //assert(PopCount(insn_ret :: io.exception :: Nil) <= 1, "these conditions must be mutually exclusive")
```

Comment:
Seems like a good check to have?

Reply:
I was having an issue compiling with gem5 because of this line suddenly. It wanted a function called sc_time_stamp() to be implemented somewhere in the c++. I wasn't sure where to put it and wanted to make sure all the csr stuff was working before i went back to figure this out.

Reply:
ok.